### PR TITLE
Run the static analysis outside of Docker

### PR DESCRIPTION
**Pull request by frankieroberto**
Hoping this will speed things up.

### CODE REVIEW

**Comment by baphled** [comment]
We're assuming that these commands will successfully run, which will not always be the case 🤔

**Comment by frankieroberto** [comment]
@baphled if they don't run, they should exit with a non-zero exit code, which will get flagged as a fail, I think?

File: infrastructure/ci/static-analysis-if-changed.sh [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT yarn eslint app config
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT yarn sass-lint -vq 'app/**/*.scss'
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT bin/brakeman --no-pager
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml down
-else
-    echo 'Static analysis not required.'
-fi
+cd cosmetics-web
+
+bundle install
```
Comment (baphled) [inline infrastructure/ci/static-analysis-if-changed.sh]: What happens when this fails? 🤔

File: infrastructure/ci/static-analysis-if-changed.sh [hunk 2]
```diff
@@ -1,17 +1,15 @@
...
+bundle exec rubocop
+bundle exec slim-lint app
+
+yarn install
```
Comment (baphled) [inline infrastructure/ci/static-analysis-if-changed.sh]: Not sure we need to do a `yarn install` 🤷‍♂

Comment (baphled) [reply infrastructure/ci/static-analysis-if-changed.sh]: If so, don't we need to make sure it's been successfully executed before the below commands have been run?

File: infrastructure/ci/static-analysis-if-changed.sh [hunk 3]
```diff
@@ -1,17 +1,15 @@
...
+yarn eslint app config
+yarn sass-lint -c .sasslint.yml -vq 'app/**/*.scss'
+
+bundle exec brakeman --no-pager
+echo "Static analysis complete - no issues found"
```
Comment (baphled) [inline infrastructure/ci/static-analysis-if-changed.sh]: Won't this hide the failing commands preceding this?

File: infrastructure/ci/static-analysis-if-changed.sh [hunk 4]
```diff
@@ -1,17 +1,15 @@
 #!/usr/bin/env bash
 set -ex
```
Comment (baphled) [inline infrastructure/ci/static-analysis-if-changed.sh]: Okay, `set -ex` will help fail fast. I should have spotted that.

I'm still not sure we need to run `bundle install` or `yarn install` though.

All deps _should_ be installed and in `Gemfile.lock` and `packages.json` when we push changes.

File: infrastructure/ci/static-analysis-if-changed.sh [hunk 5]
```diff
@@ -1,17 +1,15 @@
...
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT bin/rubocop
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT bin/slim-lint app
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT yarn eslint app config
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT yarn sass-lint -vq 'app/**/*.scss'
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml exec $COMPONENT bin/brakeman --no-pager
-    docker-compose -f docker-compose.yml -f docker-compose.ci.yml down
-else
-    echo 'Static analysis not required.'
-fi
+cd cosmetics-web
```
Comment (slorek) [inline infrastructure/ci/static-analysis-if-changed.sh]: Use `$COMPONENT` here? Unless we want to rip this out.

File: infrastructure/ci/static-analysis-if-changed.sh (hunk 1 above)
Comment (slorek) [reply infrastructure/ci/static-analysis-if-changed.sh]: Maybe `bundle check` would be preferable?

File: infrastructure/ci/static-analysis-if-changed.sh (hunk 2 above)
Comment (slorek) [reply infrastructure/ci/static-analysis-if-changed.sh]: Probably not necessary to run `yarn install`.

File: infrastructure/ci/static-analysis-if-changed.sh (hunk 3 above)
Comment (slorek) [reply infrastructure/ci/static-analysis-if-changed.sh]: Should exit before this if any preceding tests fail.

File: infrastructure/ci/static-analysis-if-changed.sh (hunk 4 above)
Comment (frankieroberto) [reply infrastructure/ci/static-analysis-if-changed.sh]: Good point. I forgot that that is part of the build process which runs before this script.

File: infrastructure/ci/static-analysis-if-changed.sh (hunk 5 above)
Comment (frankieroberto) [reply infrastructure/ci/static-analysis-if-changed.sh]: I've removed this now.

**Comment by frankieroberto** [comment]
@slorek @baphled thanks for your comments. I've made a few more changes.

`bundle install` and `yarn install` are required in order to install the gems rubocop, brakeman, slim-lint and the node modules eslint and sasslint. Previously these were installed within the docker container, but because the static analysis tests are no longer using that, they have to be installed separately.

We could install these directly (eg via `gem install rubocop` and `npm install eslint --global`) which would be faster, as there would be a lot less to install, but might be harder to maintain. I think using bundler and yarn is probably simpler.  I have removed these install commands from the `static-analysis.sh` and into a separate install step within the Travis config. This has the advantage of minimising the install steps within the report on Travis, and also makes it quicker to run the `static-analysis.sh` script locally.

So far, these improvements have resulted in the static analysis job on Travis taking around ~4 mins instead of ~10 mins.

**Pull request merged**